echo_server.py

```diff
@@ -40,46 +63,106 @@ def parse_request(header):
     except IndexError:
         pass
 
-    if method != 'GET' or proto != 'HTTP/1.1':
+    if len(first_line) < 3:
+        # For requests that are invalid due missing information:
+        raise ValueError(400, 'Bad Request')
+    # When .accept() is interupted to kill the server, method will not have
+    # been bound at the point the script continues after .accept() is forced
+    # to return, UnboundLocalError needs to be handled where parse_request()
+    # is being called.
+    elif method != 'GET' or proto != 'HTTP/1.1':
         # Deny non-GET and non-HTTP/1.1 requests.
         raise ValueError(403, 'Forbidden')
 
-    # Divide headers by line
-    try:
-        headers = []
-        for line in header[1:]:
-            headers.append(line)
-    except IndexError:
-        pass
-    return uri, headers
+    return uri
 
 
-def response_ok():
+def response_ok(content_type=None, body=None):
     response = """\
 HTTP/1.1 200 OK\r\n\
-"""
+Date: {date}\r\n\
+Content-type: {content_type}\r\n\
+Content-length: {content_length}\r\n\
+\r\n\
+{body}\r\n\
```

Comment:
You've still got the problem that you'll have too many `\n` chars in between your lines.  a browser may take it, but it ain't right.